### ISSUE TITLE
Use region set in environment variables when contacting sts with AssumeRoleWebIdentity

### ISSUE THREAD

**Issue by pkruesi** [aws/aws-sdk-php#2658]
### Describe the bug

I run some php code inside a container on eks. When I try to get credentials using `$provider = CredentialProvider::defaultProvider();` the default endpoint `sts.amazonaws.com` gets contacted instead of the regional endpoint. 

### Expected Behavior

I expect, that the function `defaultProvider()` when run in a container on eks reads the environment variables `AWS_REGION` and `AWS_DEFAULT_REGION` and uses this variables to determine which region should get used to contact sts to get credentials. For example if `AWS_DEFAULT_REGION=eu-central-1`  is set, the endpoint `sts.eu-central-1.amazonaws.com` should get conacted.

### Current Behavior

The `defaultProvider()` function uses always `us-east-1` when used with `AssumeRoleWithWebIdentity`. Except when a config array with explicit defined region is given to the `defaultProvider($config)` function.

### Reproduction Steps

Run in a container on esk with the following environment varialbes set:
```
AWS_ROLE_ARN=arn:aws:iam:xxxxxxxxxxxxxxxx:role/nexiot-api
AWS_WEB_IDENTITY_TOKEN_FILE=/var/run/secrets/eks.amazonaws.com/serviceaccount/token
AWS_STS_REGIONAL_ENDPOINTS=regional
AWS_DEFAULT_REGION=eu-central-1
AWS_REGION=eu-central-1
```
```
<?php

require_once __DIR__ . '/libraries/autoload.php';

use Aws\Credentials\CredentialProvider;
use Aws\Credentials\Credentials;
use Aws\S3\S3Client; 

$provider = CredentialProvider::defaultProvider();

$client = new S3Client([
	'region' => 'eu-central-1',
	'version' => '2006-03-01',
	'credentials' => $provider
]);

$buckets = $client->listBuckets();
```

checked with cloudtrail

```
SELECT eventTime, userIdentity.principalid, eventName, userAgent, awsRegion, eventSource  FROM xxxxxxxx-xxxx-xxxx-xxxx-xxxxxxxxxxxx 
WHERE userAgent like 'aws-sdk-php%'
	and userAgent not like 'aws-sdk-php%ubuntu%'
    and eventTime > '2023-03-09 08:00:00'
```

cloudtrail result (see awsRegion and eventSource)

```
[{"eventTime":"2023-03-09 12:16:27.000","principalid":"arn:aws:iam::xxxxxxxxxxxx:oidc-provider/oidc.eks.eu-central-1.amazonaws.com/id/XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX:sts.amazonaws.com:system:serviceaccount:xxxxxxxxxx:XXXXXXXXXXXX","eventName":"AssumeRoleWithWebIdentity","userAgent":"aws-sdk-php/3.261.0 OS/Linux/5.10.157 lang/php/7.4.26 GuzzleHttp/6.5.5 curl/7.79.1 PHP/7.4.26","awsRegion":"us-east-1","eventSource":"sts.amazonaws.com"}]
```

### Possible Solution

Evaluate the environment variables for REGION settings in the `__construct(array $config = [])` from `Credentials/AssumeRoleWithWebIdentityCredentialProvider.php`.

Will do a pull request.

### SDK version used

3.261.0

### Environment details (Version of PHP (`php -v`)? OS name and version, etc.)

PHP 7.4.26

**Comment by stobrien89** [comment]
Hi @pkruesi,

Sorry to hear about the issues.  

I think the best solution available at the moment is to use the `AWS_STS_REGIONAL_ENDPOINTS` environment variable.  The default value that the SDK resolves is `legacy`, which means that by default, the majority of STS calls will be made to their global endpoint.  Setting this value to `regional` (in either config or env var) will send requests to a regionalized endpoint.  We have some documentation on this [here](https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html).

We haven't supported `AWS_REGION` in this major version of the SDK, but you could place a `getenv` or `$ENV` into the region config to use that value.  The `AWS_DEFAULT_REGION` is specific to the AWS CLI and is something we will not consider adding support for at this time.  Hope this helps!

**Comment by pkruesi** [comment]
Hi @stobrien89 

Thank you for your answer.

> I think the best solution available at the moment is to use the `AWS_STS_REGIONAL_ENDPOINTS` environment variable. The default value that the SDK resolves is `legacy`, which means that by default, the majority of STS calls will be made to their global endpoint. Setting this value to `regional` (in either config or env var) will send requests to a regionalized endpoint. We have some documentation on this [here](https://docs.aws.amazon.com/sdkref/latest/guide/feature-sts-regionalized-endpoints.html).

I also red about the `AWS_STS_REGIONAL_ENDPOINTS` and as you can see in the snippet from the Reproduction Steps above I also posted which env variables are set in my container on eks. The `AWS_STS_REGIONAL_ENDPOINTS` variable is set to `regional` but the request is still going to `us-east-1`
Do you know why `AWS_STS_REGIONAL_ENDPOINTS` is not evaluated? Can you dig into this? Or can you point me in the right direction?

**Comment by alex-berger** [comment]
@stobrien89 Can you please elaborate a bit more, why the proposed PR https://github.com/aws/aws-sdk-php/pull/2659 is not suitable respectively what an acceptable solution should look like. 

Without more information, I can only assume it is related to the (unbelievable) mess AWS created by not automatically enabling regional STS in all regions https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_enable-regions.html. But even it the light of that, if a user explicitly sets `AWS_STS_REGIONAL_ENDPOINTS` and `AWS_REGION` (or `AWS_DEFAULT_REGION`), I see not reason to not respect those settings (as other AWS SDKs do).

**Comment by stobrien89** [comment]
Hi @pkruesi and @alex-berger,

Apologies for missing the env var you posted.  I can see now that since you're calling the default provider separately from your client, there is no config (and you're not providing it), which is why the region keeps resolving to us-east-1. 

You have a few different options here: The first would be to pass an array of config options to the default provider.  You could pass something like `'region' => getenv(AWS_REGION | AWS_DEFAULT_REGION)` and you can even pass a separately configured `sts_client`.  You could also just not provide a `'credentials'` option in your client config and a default provider will automatically be instantiated for you with the client's region and other config option.  You do lose some degree of control over the configuration this way though.

@alex-berger, we have other options available, like mentioned above. Like I mentioned in the PR, `AWS_DEFAULT_REGION` is CLI-specific, AWS_REGION is something we'd want to support at the client configuration level and we need to be careful about changing the default behavior (one of the reasons our default is to use the STS global endpoints).  While it suits this particular use case, it could break other customers that are relying on the current defaults.  We'd be open to creating some sort of config opt-in if the above suggestions don't satisfy this use case.

**Comment by alex-berger** [comment]
@stobrien89

Currently, we solve (work-around) this issue by explicitly passing in config options (as you suggested). However, we think the current implementation violates the principle of "least surprise" as it completely ignores `AWS_REGION` while other AWS SDKs respect it (e.g. Rust, Go, ...).

>  You could also just not provide a 'credentials' option in your client config and a default provider will automatically be called for you with the client's region and other config options. You do lose some degree of control over the configuration this way though. 

This is exactly, what is not working as expected. When passing no config options, it does not automatically deduce the region from `AWS_REGION`, instead it will fallback to hard-coded `us-east-1`. Here the relevant lines:
- https://github.com/aws/aws-sdk-php/blob/master/src/Credentials/CredentialProvider.php#L471
- https://github.com/aws/aws-sdk-php/blob/master/src/Credentials/AssumeRoleWithWebIdentityCredentialProvider.php#L76

**Comment by pkruesi** [comment]
@stobrien89 

> Apologies for missing the env var you posted. I can see now that since you're calling the default provider separately from your client, there is no config (and you're not providing it), which is why the region keeps resolving to us-east-1.

I am still not understand why `AWS_STS_REGIONAL_ENDPOINTS=regional` is not evaluated. Why should a `defaultProvider()` resolve to a different region if `AWS_STS_REGIONAL_ENDPOINTS=regional` is set. Its quiet confusing.

**Comment by stobrien89** [comment]
@alex-berger,

I agree that we should have more parity with other AWS SDKs here in terms of utilizing `AWS_REGION`.  I'm not certain we want this evaluated in the provider we're discussing here though.  Can you explain your use case a bit further? I'm assuming you want to configure a separate region than the one you've configured your client with? If so, why?

@pkruesi,

From what I can tell, the environment variable is being evaluated correctly.  It seems the problem in your case is that the region always falls back to `us-east-1` because there is no region being provided in the config when you call the default provider separately from your client.

**Comment by alex-berger** [comment]
> I'm assuming you want to configure a separate region than the one you've configured your client with? If so, why?

@stobrien89 No, actually all we want is, that it respects `AWS_REGION` and thus uses the **same** region as our clients. The problem is, that it silently default to `us-east-1` instead of using `AWS_REGION`.

**Comment by stobrien89** [comment]
@alex-berger,

In that case you can just omit the `'credentials'` option from your client config.  The default credential provider will be called using your client config (including the same region your client uses).  This is what we typically recommend unless you have a use case that requires you to use a different region/other options than your client.

**Comment by bruecktech** [comment]
Yesterdays us-east-1 outage caused unexpected issues for our services running in eu-central-1 because of this wrong behaviour.
We are also setting `AWS_REGION=eu-central-1` but still the `AssumeRoleWithWebIdentity` provider silently calls sts in us-east-1. `AWS_STS_REGIONAL_ENDPOINTS=regional` is also configured in our case but it's ignored.
Can you please re-open this issue and get it fixed?

> In that case you can just omit the 'credentials' option from your client config. The default credential provider will be called using your client config (including the same region your client uses)

This explanation is exactly what you would expect, but as described above it's not the case.

**Comment by stobrien89** [comment]
Hi @bruecktech, all,

Apologies for the delayed response.  Comments on closed issues are sometimes hard for us to see, and #3197 had been opened, which caused a bit of confusion with the support case that was opened for this.  We initially wanted to restore our legacy behavior, which was to fall back to `us-east-1` calling the global STS endpoint, but we've arrived at making every reasonable attempt to derive a region before falling back. This was merged last night in https://github.com/aws/aws-sdk-php/pull/3203 and is available in version [3.358.0](https://github.com/aws/aws-sdk-php/releases/tag/3.358.0).

Now `AssumeRoleWithWebIdentityCredentialProvider` will look for a region in the config passed to it— This will be your client's region if `AssumeRoleWithWebIdentityCredentialProvider` is created by the client in the default chain, or if you're creating `AssumeRoleWithWebIdentityCredentialProvider` in your config, it would be the equivalent of passing `'region' => '<your_region>'`.  If it can't derive a region that way, then it falls back to `AWS_REGION`, if available, then `us-east-1` with a notice level error that you should probably be configuring a region.  I'll post any more updates I receive, but let us know if you have any other questions.
